### PR TITLE
set projection order when project persistent buffer to its persistent producers

### PR DESCRIPTION
**Pull request by liqiangxl**
Issue: #1406
Fix: set a projection order based on dependencies when project persistent buffer to its persistent producers.
Example:
```
      // visit consumer before producer. e.g.
      // T1 = f(T0); Tx = add(T1, broadcast(sum(T1)));
      // T2 = f(T1); Ty = add(T2, broadcast(sum(T2)));
      // T3 = f(T2); Tz = add(T3, broadcast(sum(T3)));
      // T1, T2, T3 are persistent buffers.
      // The visiting order should be [T3, T2, T1].
      // After project T3 to its producers, we have:
      // Tz = add(f(T2),broadcast(sum(T3)));
      // After project T2 to its producers, we have:
      // Tz = add(f(f(T1)),broadcast(sum(T3)));
      // Ty = add(f(T1), broadcast(sum(T2)));
      // At last, the only persistent buffer is T1.
      // For a solid case, see NVFuserTest.ChainProjectionToPersistentProducer.
```

### CODE REVIEW

**Comment by liqiangxl** [comment]
!build --diff --diff-bench

**Comment by liqiangxl** [comment]
The CI failures are not related to the changes in this PR.

File: csrc/scheduler/reduction_utils.cpp [hunk 1]
```diff
@@ -736,18 +736,38 @@ class PersistentBufferProjector {
...
+      for (auto buffer_i : visiting_order) {
         auto buffer = persistent_buffers[buffer_i];
         const auto& producers = ir_utils::producerTvsOf(buffer);
         if (scheduler_utils::canProjectToPersistentProducer(
                 buffer, producers, persistent_buffer_set)) {
           projectToInputOrImmediatePersistentProducer(
               (int)buffer_i,
               std::vector<Val*>(producers.begin(), producers.end()));
-          // "buffer" is no longer a persistent buffer
-          persistent_buffer_set.erase(buffer);
```
Comment (naoyam) [inline csrc/scheduler/reduction_utils.cpp:750]: Why is this removed?

File: csrc/scheduler/reduction_utils.cpp [hunk 2]
```diff
@@ -736,18 +736,38 @@ class PersistentBufferProjector {
             (int)buffer_i, fusion_->inputs());
       }
     } else {
+      // visit consumer before producer. e.g.
```
Comment (naoyam) [inline csrc/scheduler/reduction_utils.cpp:739]: This `project` function is getting bigger. Can you do some cleanup? Just separating it to `projectToInputs()` and `projectToProducers` would be enough.

File: csrc/scheduler/reduction_utils.cpp [hunk 3]
```diff
@@ -736,18 +736,38 @@ class PersistentBufferProjector {
...
+      // After project T3 to its producers, we have:
+      // Tz = add(f(T2),broadcast(sum(T3)));
+      // After project T2 to its producers, we have:
+      // Tz = add(f(f(T1)),broadcast(sum(T3)));
+      // Ty = add(f(T1), broadcast(sum(T2)));
+      // At last, the only persistent buffer is T1.
+      // For a solid case, see NVFuserTest.ChainProjectionToPersistentProducer.
+      std::vector<int> visiting_order(persistent_buffers.size());
+      std::iota(visiting_order.begin(), visiting_order.end(), 0);
+      std::sort(
```
Comment (naoyam) [inline csrc/scheduler/reduction_utils.cpp:754]: Couldn't this be a source of non-determinism? Shouldn't we use `std::stable_sort`?

File: csrc/scheduler/reduction_utils.cpp (hunk 1 above)
Comment (liqiangxl) [reply csrc/scheduler/reduction_utils.cpp:750]: Once a "buffer" has been projected, it is no longer a persistent buffer. Nonetheless, its removal from the buffer set is unnecessary due to the established visiting order. This order guarantees that buffers accessed prior to the current one cannot be the producer of the current buffer. Therefore, additional cleanup is not required.

File: csrc/scheduler/reduction_utils.cpp (hunk 2 above)
Comment (liqiangxl) [reply csrc/scheduler/reduction_utils.cpp]: changed.

File: csrc/scheduler/reduction_utils.cpp (hunk 3 above)
Comment (liqiangxl) [reply csrc/scheduler/reduction_utils.cpp]: changed to `std::stable_sort`

**Review by naoyam — APPROVED** [review]
LGTM
